src/flash.cu

```diff
@@ -1,384 +1,322 @@
-                    			   int head_dim,
-                    			   int q_batch_stride,
-                    			   int k_batch_stride,
-                    			   int v_batch_stride,
-                    			   int o_batch_stride,
-                    			   int q_head_stride,
-                    			   int k_head_stride,
-                    			   int v_head_stride,
-                    			   int o_head_stride,
-                    			   int q_row_stride,
-                    			   int k_row_stride,
-                    			   int v_row_stride,
-                    			   int o_row_stride,
-					   int num_blocks,
-                                           int block_size,
-					   int max_num_blocks_per_seq,
-					   int block_table_batch_stride,
-                    			   float softmax_scale,
-                    			   bool is_causal,
-                    			   int window_size_left,
-                    			   int window_size_right,
-					   int num_splits,
-                    			   cudaStream_t stream) {
-  const int seqlen_k = max_num_blocks_per_seq * block_size;
-
-  auto params = get_fwd_params(q_ptr, kcache_ptr, vcache_ptr, output_ptr,
-			       batch_size, seqlen_q, seqlen_k,
-			       num_heads, num_heads_k, head_dim,
-			       q_batch_stride, k_batch_stride,
-			       v_batch_stride, o_batch_stride,
-			       q_head_stride, k_head_stride,
-			       v_head_stride, o_head_stride,
-			       q_row_stride, k_row_stride,
-			       v_row_stride, o_row_stride,
-			       softmax_scale, is_causal,
-			       window_size_left, window_size_right);
-
-  set_splitkv_params(params, block_table_ptr, seqlens_k_ptr, softmax_lse_accum_ptr,
-                     output_accum_ptr, batch_size, seqlen_q, seqlen_k, num_heads,
-                     head_dim, num_blocks, block_size, max_num_blocks_per_seq,
-		     block_table_batch_stride, num_splits);
-
-  run_splitkv(params, stream);
-}
-
-} // namespace flash_attn
\ No newline at end of file
+#include <cuda.h>
+
+#include <cstdio>
+#include <stdexcept>
+#include <string>
+
+#include "../hopper/flash_fwd_launch_template.h"
+#include "flash.h"
+#include "flash_internal.h"
+
+namespace flash_attn {
+
+inline void _assert(bool result, const char* const file, int const line,
+                    std::string const& info = "") {
+  if (!result) {
+    throw std::runtime_error(std::string("[ERROR] ") + info + " Assertion fail: " + file + ":" +
+                             std::to_string(line) + " \n");
+  }
+}
+
+#define CHECK(val) _assert(val, __FILE__, __LINE__)
+
+void run(Flash_fwd_params params, cudaStream_t stream) {
+  int device;
+  cudaGetDevice(&device);
+  int major, minor;
+  cudaDeviceGetAttribute(&major, cudaDevAttrComputeCapabilityMajor, device);
+  cudaDeviceGetAttribute(&minor, cudaDevAttrComputeCapabilityMinor, device);
+  params.sm = major * 10 + minor;
+
+  if(params.sm >= 90 && (params.d == 64 || params.d == 128 || params.d == 256)) {
+    if (!params.is_e4m3) {

```

Comment:
Change this to `if (params.is_e4m3)` and put `cutlass::float_e4m3_t` specializations here. This is cleaner.